compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,644 @@
+            case TYPE_REFERENCE -> checkTypeAndDetectDuplicateFields(
+                    ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor(), ctx);
+            case INTERSECTION -> checkTypeAndDetectDuplicateFields(getRawType(typeSymbol), ctx);
+        }
+    }
+
+    private FunctionCallExpressionNode getFunctionCallExpressionNodeIfPresent(ExpressionNode expressionNode) {
+        return switch(expressionNode.kind()) {
+            case CHECK_EXPRESSION -> getFunctionCallExpressionNodeIfPresent(
+                    ((CheckExpressionNode) expressionNode).expression());
+            case FUNCTION_CALL -> (FunctionCallExpressionNode) expressionNode;
+            default -> null;
+        };
+    }
+
+    private Optional<String> getFunctionName(FunctionCallExpressionNode node) {
+        NameReferenceNode nameReferenceNode = node.functionName();
+        if (nameReferenceNode.kind() != SyntaxKind.QUALIFIED_NAME_REFERENCE) {
+            return Optional.empty();
+        }
+        QualifiedNameReferenceNode qualifiedNameReferenceNode = (QualifiedNameReferenceNode) nameReferenceNode;
+        String prefix = qualifiedNameReferenceNode.modulePrefix().text();
+        if (!prefix.equals(modulePrefix)) {
+            return Optional.empty();
+        }
+        return Optional.of(qualifiedNameReferenceNode.identifier().text());
+    }
+
+    private boolean isParseFunctionOfStringSource(ExpressionNode expressionNode) {
+        FunctionCallExpressionNode node = getFunctionCallExpressionNodeIfPresent(expressionNode);
+        if (node == null) {
+            return false;
+        }
+        Optional<String> functionName = getFunctionName(node);
+        return functionName
+                .map(fn -> fn.contains(Constants.PARSE_STRING) || fn.contains(Constants.PARSE_BYTES) ||
+                        fn.contains(Constants.PARSE_STREAM) || fn.contains(Constants.TRANSFORM) ||
+                        fn.contains(Constants.PARSE_LISTS))
+                .orElse(false);
+    }
+
+    private void validateFunctionParameterTypes(ExpressionNode expressionNode,
+                                                TypeSymbol expType,
+                                                Location currentLocation, SyntaxNodeAnalysisContext ctx) {
+        FunctionCallExpressionNode node = getFunctionCallExpressionNodeIfPresent(expressionNode);
+        if (node == null) {
+            return;
+        }
+        Optional<String> functionName = getFunctionName(node);
+        SeparatedNodeList<FunctionArgumentNode> args = node.arguments();
+        functionName.ifPresent(fn ->
+                validateFunctionParameterTypesWithExpType(expType, currentLocation, ctx, fn, args));
+    }
+
+    private void validateFunctionParameterTypesWithExpType(TypeSymbol expType, Location currentLocation,
+                                                           SyntaxNodeAnalysisContext ctx, String functionName,
+           SeparatedNodeList<FunctionArgumentNode> args) {
```

Comment:
Formatting here seems bit off
```suggestion
    private void validateFunctionParameterTypesWithExpType(TypeSymbol expType, Location currentLocation,
                                                           SyntaxNodeAnalysisContext ctx, String functionName,
                                                           SeparatedNodeList<FunctionArgumentNode> args) {
```